stockroom/storages/model.py

```diff
@@ -0,0 +1,120 @@
+    specifically to store the shape of each layer. This shape arrayset is needed
+    because the weights of each layer would be flattened before saving. This is
+    essential since handling variable shapes and variable ranks are more complex
+    than flattening and reshaping-back the weights.
+    """
+
+    def __init__(self, write):
+        self.co = None
+        self._write = write
+        self._repo = StockRepository()
+
+    def save(self, name, model):
+        if not self._write:
+            raise PermissionError("ModelStore instance is not write-enabled")
+        if hasattr(model, 'state_dict'):
+            state = model.state_dict()
+            weights = [x.numpy() for x in state.values()]
+            layers = state.keys()
+            library = 'torch'
+            library_version = torch.__version__
+        elif hasattr(model, 'get_weights'):
+            library = 'tf'
+            layers = None
+            weights = model.get_weights()
+            library_version = tf.__version__
+        else:
+            raise TypeError("Unknown model type. StockRoom can work with only "
+                            "``Keras.Model`` or ``torch.nn.Module`` modules")
+        longest = max([len(x.reshape(-1)) for x in weights])
+        dtypes = [w.dtype.name for w in weights]
+
+        co = self._repo.checkout(write=True)
+        co.metadata[parser.model_metakey(name, 'library')] = library
+        co.metadata[parser.model_metakey(name, 'libraryVersion')] = library_version
+        co.metadata[parser.model_metakey(name, 'longest')] = str(longest)
+        co.metadata[parser.model_metakey(name, 'dtypes')] = parser.stringify(dtypes)
+        co.metadata[parser.model_metakey(name, 'numLayers')] = str(len(weights))
+        co.metadata[parser.model_metakey(name, 'layers')] = parser.stringify(layers)
+
+        # ---------- Create arraysets if not exist -----------------
+
+        shapeKey = parser.shapekey(name, str(longest))
+        if shapeKey not in co.arraysets.keys():
+            co.arraysets.init_arrayset(shapeKey, 10, np.int64, variable_shape=True)
+        for i, w in enumerate(weights):
+            modelKey = parser.modelkey(name, str(longest), dtypes[i])
+            if modelKey not in co.arraysets.keys():
+                co.arraysets.init_arrayset(
+                    modelKey, longest,np.dtype(dtypes[i]), variable_shape=True)
+
+        # ---------------------------------------------------------
+
+        with co:
+            shape_aset = co.arraysets[shapeKey]
+            for i, w in enumerate(weights):
+                model_aset = co.arraysets[parser.modelkey(name, longest, dtypes[i])]
+                model_aset[i] = w.reshape(-1)
+                if w.shape:
+                    shape_aset[i] = np.array(w.shape)
+                else:
+                    shape_aset[i] = np.array(()).astype('int64')
+        co.close()
+
+    def load(self, name, model):
+        co = self._repo.checkout()
```

Comment:
use a context manager. if any error occurs below then the checkout will not be closed.

Reply:
handeld automatically if https://github.com/tensorwerk/stockroom/pull/1#discussion_r361956040 is implemented.